### PR TITLE
Check out PRs from a fork

### LINKED ISSUE #3395: Clicking on a PR from a fork does nothing
https://github.com/desktop/desktop/blob/1de8b496393a316b23d4492adf6aa9df9f44bd57/app/src/ui/branches/index.tsx#L215-L219

### PR DESCRIPTION
**Pull request by joshaber**
Fixes #3395

### Cleaning up
- [x] Initiated when PRs are refreshed
- [x] Iterate over remotes
- [x] Delete any remote, using our special magic prefix, not referenced by an open PR 
- [x] Reload current remote

### Clicking a PR
- [x] Add a remote with a magic-prefix if needed 
- [x]  Fetch (this will prune unreferenced refs for us)
- [x] Checkout ref as `pr/[PR Number]`

### CODE REVIEW

**Comment by iAmWillShepherd** [comment]
We've run into a bug where we are stripping out valid characters from a branch name. This is preventing us from using our special `github-desktop` magic prefix.

**Comment by timofonic** [comment]
@joshaber @iAmWillShepherd Will this be possible to get fixed and merged someday? Please... 
Thanks in advance :D

**Comment by iAmWillShepherd** [comment]
Yes. Most of the team is out for the holidays, including me, so progress will be a bit slow, but it will happen! 🎅

**Comment by iAmWillShepherd** [comment]
From @nerdneha 

1. Have another member (or evil account, etc.) create a fork off of an existing repository and modify a file
2. Have that member commit and open a pull request into the base fork via GitHub
3. Click on PR in GitHub desktop to check it out, verify that the remote was added to the forked PR
4. Check out that pull request on the desktop application
5. Verify that a branch has been opened in the desktop application
6. Verify that the (remote?/pull request?) exists on the machine upon refresh with the prefix `github-desktop-*?`  (`git remove -v`)
7. Close the pull request via GitHub
8. Refresh on the desktop application
9. Verify that the (pull request?/remote??) is no longer on the local machine (`git remove -v`)
10. Verify that the branch exists on the desktop application but not the pull request
11. Successfully delete the branch without issue (

**Comment by nerdneha** [comment]
tl;dr we proved that pruning closed pull requests works correctly!

If you follow the steps to a tee, the checking out AND pruning features work. Essentially you are creating a remote pull request and are able to successfully fetch it and see it on the desktop application. Then, after closing the pull request, you will see that your machine has removed the remote. This is important in order to prevent having a flood of remotes that are no longer pertinent on your machine that never disappear. This does not handle the case that you might have millions of open PRs that are also on your machine and are not pruned.

One edge case we discovered is if the base-fork member/owner creates the pull-requests, the remotes will not show in the machine via `git remote -v` and therefore will not need to be pruned. This is working as expected.

**Comment by joshaber** [comment]
>One edge case we discovered is if the base-fork member/owner creates the pull-requests, the remotes will not show in the machine via git remote -v and therefore will not need to be pruned. This is working as expected.

To be clear, this is because the pull request came from within the repository itself, so there was no need to add another remote. The remote already exists. Do I have that right?

**Comment by iAmWillShepherd** [comment]
@joshaber, that is correct. I think this PR may be a good use case for unit tests.

File: app/src/lib/stores/pull-request-store.ts [hunk 1]
```diff
@@ -14,6 +14,16 @@ import {
...
+import { Repository } from '../../models/repository'
+import { getRemotes, removeRemote } from '../git'
+import { IRemote } from '../../models/remote'
+
+/**
+ * This is the magic remote name prefix
+ * for when we add a remote on behalf of
+ * the user.
+ */
+export const ForkedRemotePrefix = 'github-desktop-'
```
Comment (shiftkey) [inline app/src/lib/stores/pull-request-store.ts]: :memo: Could we incorporate some public-facing docs into this PR? It'd be nice to summarize what we're adding to a repository in terms of remotes and branch prefixes, so we can easily refer people with questions and issues to it.

File: app/src/lib/stores/app-store.ts [hunk 2]
```diff
@@ -3008,8 +3010,9 @@ export class AppStore {
       return
     }
 
-    await this.pullRequestStore.refreshPullRequests(gitHubRepository, account)
-    return this.updateMenuItemLabels(repository)
+    await this.pullRequestStore.refreshPullRequests(repository, account)
+
+    this.updateMenuItemLabels(repository)
```
[the remark this answers is not in the document]
Comment (iAmWillShepherd) [reply app/src/lib/stores/app-store.ts]: The result of that is `void` and the function this lives in returns a `Promise<void>`.

**Comment by iAmWillShepherd** [comment]
🔢

**Comment by iAmWillShepherd** [comment]
@shiftkey, I'm going to remove the `ready-for-review` label until docs are complete. But Feel free to continue reviewing the code portions.

File: docs/technical/pull-requests.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Checking out pull requests from a forked repository
+PR [#3602](https://github.com/desktop/desktop/pull/3602) introduced the ability to checkout a branch from a forked repository. In order to accomplish this, we needed a way to manage remotes on your behalf. This document is intended to detail the process we developed to make checking out PRs as frictionless as possible.
+
+## Magic Remote Prefix
+One of the main problems we needed to solve was determining which remotes are no longer needed and can be cleaned. We decided to prefix the remotes we add on your behalf with a magic string: `github-desktop-`
```
Comment (nerdneha) [inline docs/technical/pull-requests.md]: github-desktop-[username]

File: docs/technical/pull-requests.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+```ts
+export const ForkedRemotePrefix = 'github-desktop-'
+```
+[Code](https://github.com/desktop/desktop/blob/34a05b155ff69bb19cc4da5b2caa89856e3e63fb/app/src/lib/stores/pull-request-store.ts#L26)
+
+## Cleaning Remotes
+One of the goals of our design was to ensure that we don’t cause your remotes — `.git/refs/remotes` — to grow unbounded. We prevent this by cleaning up after ourselves. We determined that a remote is a candidate for removal when it meets the following conditions.
```
Comment (nerdneha) [inline docs/technical/pull-requests.md]: Maybe we can put a note that we will remove the remote but the branch will remain untouched

File: docs/technical/pull-requests.md (hunk 3 above)
Comment (nerdneha) [reply docs/technical/pull-requests.md]: I'm going back and forth with whether we say "github-desktop-" as the prefix or just say that the remote is titled "github-desktop-[username]"... feel free to ignore if you have a strong opinion

Comment (joshaber) [reply docs/technical/pull-requests.md]: Could we swap "can be cleaned" with "can be removed"? "Cleaned" may not be clear as it's not git's terminology.

File: docs/technical/pull-requests.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+## Magic Remote Prefix
+One of the main problems we needed to solve was determining which remotes are no longer needed and can be cleaned. We decided to prefix the remotes we add on your behalf with a magic string: `github-desktop-`
+
+```ts
+export const ForkedRemotePrefix = 'github-desktop-'
+```
+[Code](https://github.com/desktop/desktop/blob/34a05b155ff69bb19cc4da5b2caa89856e3e63fb/app/src/lib/stores/pull-request-store.ts#L26)
+
+## Cleaning Remotes
```
Comment (joshaber) [inline docs/technical/pull-requests.md]: I think this section should be moved up, since it's defining the problem that we're trying to solve with the magic prefix.

File: docs/technical/pull-requests.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+One of the goals of our design was to ensure that we don’t cause your remotes — `.git/refs/remotes` — to grow unbounded. We prevent this by cleaning up after ourselves. We determined that a remote is a candidate for removal when it meets the following conditions.
+
+* Start with our prefix
+* The PR associated with the remote is closed
+
+```ts
+private forkedRemotesToDelete(
```
Comment (joshaber) [inline docs/technical/pull-requests.md]: I worry that including a c&p of the code here means it'll fall out of sync with the _actual_ code we use. Could we describe the algorithm in some pseudo code, so we don't get bogged down in the implementation details?

File: app/src/lib/stores/app-store.ts [hunk 7]
```diff
@@ -3153,4 +3151,69 @@ export class AppStore {
...
+      repository.gitHubRepository
+    )
+    const head = pullRequest.head
+    const isRefInThisRepo =
+      head.gitHubRepository &&
+      head.gitHubRepository.cloneURL === gitHubRepository.cloneURL
+
+    if (isRefInThisRepo) {
+      // We need to fetch FIRST because someone may have created a PR since the last fetch
+      await this._fetch(repository, FetchType.UserInitiatedTask)
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: Does this need to be user initiated? I think the only real difference between the behaviour is error handling - we might want to block the subsequent checkout if this fails to get the new ref. Am I reading that right?

File: app/src/lib/stores/app-store.ts [hunk 8]
```diff
@@ -3153,4 +3151,69 @@ export class AppStore {
...
+      await this._checkoutBranch(repository, head.ref)
+    } else if (head.gitHubRepository != null) {
+      const cloneURL = forceUnwrap(
+        "This pull request's head is not populated but should be",
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: This error message should talk about the clone URL

File: app/src/lib/stores/app-store.ts [hunk 9]
```diff
@@ -3153,4 +3151,69 @@ export class AppStore {
 
     return gitStore.addUpstreamRemoteIfNeeded()
   }
+
+  private forkPullRequestRemoteName(remoteName: string) {
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: This can be a function outside the class as it doesn't depend on any internal state

File: app/src/lib/stores/app-store.ts [hunk 10]
```diff
@@ -3153,4 +3151,69 @@ export class AppStore {
...
+        head.gitHubRepository.owner.login
+      )
+      const remotes = await getRemotes(repository)
+      const remote = remotes.find(r => r.name === remoteName)
+
+      if (remote == null) {
+        await addRemote(repository, remoteName, cloneURL)
+      } else if (remote.url !== cloneURL) {
+        log.error(
+          `Expected PR remote ${remoteName} url to be ${cloneURL} got ${
```
Comment (shiftkey) [inline app/src/lib/stores/app-store.ts]: This error is logged, but we don't surface anything to the user about this mismatch. I gather the eventual `checkout` will fail because it can't find the ref - is there anything else we can do here to make this friendlier?

File: app/src/lib/stores/helpers/pull-request-updater.ts [hunk 11]
```diff
@@ -104,9 +106,14 @@ export class PullRequestUpdater {
     this.timeoutHandles.set(TimeoutHandles.PushedPullRequest, handle)
   }
 
-  private async refreshPullRequestStatus(repository: GitHubRepository) {
-    await this.store.refreshPullRequestStatuses(this.repository, this.account)
-    const prs = await this.store.getPullRequests(repository)
+  private async refreshPullRequestStatus(repository: Repository) {
```
Comment (shiftkey) [inline app/src/lib/stores/helpers/pull-request-updater.ts]: We already track `this.repository` from the constructor - we can eliminate this parameter for a bit of :lipstick:

File: app/src/lib/stores/helpers/pull-request-updater.ts [hunk 12]
```diff
@@ -80,10 +85,7 @@ export class PullRequestUpdater {
   }
 
   /** Starts fetching the statuses of PRs at an accelerated rate */
-  public didPushPullRequest(
-    repository: GitHubRepository,
-    pullRequest: PullRequest
-  ) {
+  public didPushPullRequest(repository: Repository, pullRequest: PullRequest) {
```
Comment (shiftkey) [inline app/src/lib/stores/helpers/pull-request-updater.ts]: This `repository` is also unnecessary to pass in as a parameter

File: app/src/lib/stores/helpers/pull-request-updater.ts [hunk 13]
```diff
@@ -45,6 +45,11 @@ export class PullRequestUpdater {
 
   /** Starts the updater */
   public start() {
+    const githubRepo = forceUnwrap(
+      'Can only fetch pull requests for GitHub repositories',
```
Comment (shiftkey) [inline app/src/lib/stores/helpers/pull-request-updater.ts]: :lipstick: this is about updating PR information rather than just fetching, right?

File: app/src/lib/stores/app-store.ts (hunk 7 above)
Comment (iAmWillShepherd) [reply app/src/lib/stores/app-store.ts]: Yea, that seems like a reasonable thing to do. Do you suggest making it a `BackgroundTask` task and wrapping it in a `try/catch` block?

File: app/src/lib/stores/app-store.ts (hunk 10 above)
Comment (iAmWillShepherd) [reply app/src/lib/stores/app-store.ts]: Yea, we should probably surface this to the user and return from this point. @nerdneha @joshaber you two have any different ideas?

File: app/src/lib/stores/helpers/pull-request-updater.ts (hunk 13 above)
Comment (iAmWillShepherd) [reply app/src/lib/stores/helpers/pull-request-updater.ts]: In this case, it's just about fetching so that we grab new PRs created on dotcom. I suppose this can also be seen as updating PR info (that is _why_ we do the fetch in the first place).

File: app/src/lib/stores/app-store.ts (hunk 10 above)
Comment (joshaber) [reply app/src/lib/stores/app-store.ts]: Yeah, a decent start would be posting an error so the user knows it failed.

Comment (nerdneha) [reply app/src/lib/stores/app-store.ts]: +1 on showing error but maybe not with the text that we use for logging; maybe just saying there was an error trying to access the PR?

**Comment by iAmWillShepherd** [comment]
🔢

File: app/src/lib/stores/app-store.ts (hunk 7 above)
Comment (shiftkey) [reply app/src/lib/stores/app-store.ts]: A `BackgroundTask` will not propagate the error up to the user, so a `try/catch` isn't necessary.

At this point I think there's two potential solutions:

 - use `UserInitiatedTask` and, if the fetch fails, an error gets shown and no checkout is attempted.

This is the safer option as it means you avoid the "missing ref" situation mentioned in the comment, but also ensures that you're working with the latest commit even if it's a known ref.

 - switch to `BackgroundTask` and, if the fetch fails, the code continues on and tries to checkout

This is a bit more risky, and means the actual issue (no network, so can't get new refs) is not actually mentioned in the error if it tries to checkout a missing ref.

I think we should just leave this as-is, and let it fail fast.

File: app/src/lib/stores/helpers/pull-request-updater.ts (hunk 13 above)
Comment (shiftkey) [reply app/src/lib/stores/helpers/pull-request-updater.ts]: I'm nit-picking the mixing of Git concepts and API concepts here, and for a message that's essentially for developers it's not worth bikeshedding. As long as this message is unique enough to be able to trace if we see it in the wild, I'm :thumbsup:
